KeeAgent/KeeAgentUiThread.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Threading;
+using System.Windows.Forms;
+
+namespace KeeAgent
+{
+  /// <summary>
+  /// Provides a separate UI thread for interactive communication without interlocking with the MainWindow.
+  /// </summary>
+  public sealed class KeeAgentUiThread : IDisposable
+  {
+    private const string ComponentName = "KeeAgent Interactive UI";
+    private SynchronizationContext _synchronizationContext;
+    private ApplicationContext _applicationContext;
+
+    public KeeAgentUiThread()
+    {
+      var uiThread = new Thread(UiThreadMain) {
+        Name = ComponentName,
+        IsBackground = true // active background thread will not block the application shutdown process
+      };
+      uiThread.SetApartmentState(ApartmentState.STA);
+      uiThread.Start();
+    }
+
+    private void UiThreadMain()
+    {
+      _synchronizationContext = new WindowsFormsSynchronizationContext();
+      SynchronizationContext.SetSynchronizationContext(_synchronizationContext);
+      _applicationContext = new ApplicationContext();
+      Application.Run(_applicationContext);
+    }
+
+    /// <summary>
+    /// Performs message loop shutdown.
+    /// </summary>
+    public void Dispose()
+    {
+      if (_synchronizationContext == null) {
+        return;
+      }
+
+      var synchronizationContext = _synchronizationContext;
+      _synchronizationContext = null;
+      var applicationContext = _applicationContext;
+      _applicationContext = null;
+      synchronizationContext.Post(_ => applicationContext.ExitThread(), null);
```

Comment:
And here call `Application.ExitThread()` which effectively does the same thing.

Reply:
@dlech , as for me it was a little bit more expressive with the `ApplicationContext` in contrast with not very obvious static method. Since effectively it does the same and you prefer it that way, I got rid of the redundant variable.